pkg/agent/multicast/mcast_discovery.go

```diff
@@ -97,14 +99,50 @@ func (s *IGMPSnooper) queryIGMP(group net.IP, versions []uint8) error {
 		if err != nil {
 			return err
 		}
-		if err := s.ofClient.SendIGMPQueryPacketOut(igmpQueryDstMac, mcastAllHosts, openflow13.P_NORMAL, igmp); err != nil {
+		// outPort sets the output port of the packetOut message. We expect the message to go through OVS pipeline
+		// from table0. The final OpenFlow message will use a standard OpenFlow port number OFPP_TABLE = 0xfffffff9 corrected
+		// by ofnet.
+		outPort := uint32(0)
+		if err := s.ofClient.SendIGMPQueryPacketOut(igmpQueryDstMac, types.McastAllHosts, outPort, igmp); err != nil {
 			return err
 		}
-		klog.V(2).InfoS("Sent packetOut for IGMP query", "group", group.String(), "version", version)
+		klog.V(2).InfoS("Sent packetOut for IGMP query", "group", group.String(), "version", version, "outPort", outPort)
 	}
 	return nil
 }
 
+func (s *IGMPSnooper) validate(event *mcastGroupEvent) (bool, error) {
+	if s.validator == nil {
+		// Do not need to validate if the packet should be dropped since Validator is null, thus here returns directly
+		return true, nil
+	}
+	if event.iface.Type != interfacestore.ContainerInterface {
+		return true, fmt.Errorf("interface is not container")
+	}
+	// Validates check if packet should be dropped or not, and return multicast NP information
+	item, err := s.validator.Validate(event.iface.PodName, event.iface.PodNamespace, event.group)
+	if err != nil {
+		// It shall drop the packet if function Validate returns error
+		klog.ErrorS(err, "Failed to validate multicast group event")
+		return false, err
+	}
+	if item.RuleAction != "" {
+		klog.V(2).InfoS("Got NetworkPolicy action for IGMP query", "RuleAction", item.RuleAction,
+			"uuid", item.UUID, "Name", item.Name)
+	}
+	if item.RuleAction == v1alpha1.RuleActionDrop {
+		return false, nil
+	}
+	return true, nil
+}
+
+func (s *IGMPSnooper) validatePacketAndNotify(event *mcastGroupEvent) {
+	allow, _ := s.validate(event)
```

Comment:
One question: if a Pod member already joined a group, then admin add a rule to drop the IGMP report from this Pod, Antrea Agent does not remove the Pod from the OpenFlow groups buckets directly, but it will be removed when after timeout? 

This looks a bit different from the expectation?

Reply:
Yes, the pod is still in the group unless it is removed after timeout. Do you mean remove it from the openflow group directly?

Reply:
done